src/serial/usart.rs

```diff
@@ -349,6 +366,91 @@ macro_rules! uart_shared {
             }
         }
 
+        impl<Pin> embedded_io::ErrorType for Tx<$USARTX, Pin, NoDMA> {
+            type Error = Error;
+        }
+        impl<Pin> WriteReady for Tx<$USARTX, Pin, NoDMA> {
+            fn write_ready(&mut self) -> Result<bool, Self::Error> {
+                let usart = unsafe { &(*$USARTX::ptr()) };
+                Ok(usart.isr.read().txe().bit_is_set())
+            }
+        }
+        // writes until fifo (or tdr) is full
+        impl<Pin> embedded_io::Write for Tx<$USARTX, Pin, NoDMA> {
+            fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Error> {
+                if buf.len() == 0 { return Ok(0) }
+                let usart = unsafe { &(*$USARTX::ptr()) };
+                while !self.write_ready()? {
+                    core::hint::spin_loop()
+                }
+                // can't know fifo capacity in advance
+                let count = buf.into_iter()
+                    .take_while(|_| usart.isr.read().txe().bit_is_set())
+                    .map(|b| usart.tdr.write(|w| unsafe { w.tdr().bits(*b as u16) }))
+                    .count();
+
+                Ok(count)
+            }
+            fn flush(&mut self) -> Result<(), Error> {
+                nb::block!(embedded_hal_old::serial::Write::<u8>::flush(self))
+            }
+        }
+
+        impl<Pin> embedded_io::ErrorType for Rx<$USARTX, Pin, NoDMA> {
+            type Error = Error;
+        }
+        impl<Pin> ReadReady for Rx<$USARTX, Pin, NoDMA> {
+            fn read_ready(&mut self) -> Result<bool, Self::Error> {
+                match self.data_ready() {
+                    Ok(()) => Ok(true),
+                    Err(nb::Error::WouldBlock) => Ok(false),
+                    Err(nb::Error::Other(e)) => Err(e),
+                }
+            }
+        }
+        impl<Pin> embedded_io::Read for Rx<$USARTX, Pin, NoDMA> {
+            fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error> {
+                if buf.len() == 0 { return Ok(0) }
+                let usart = unsafe { &(*$USARTX::ptr()) };
+                let mut count = 0;
+
+                while !self.read_ready()? {
+                    core::hint::spin_loop()
+                }
+                while self.read_ready()? && count < buf.len() {
+                    buf[count] = usart.rdr.read().bits() as u8;
+                    count += 1
+                }
```

Comment:
Will this ever yield more than 1 byte(assuming f_sys > baud_rate)?

Reply:
Manual says there's an 8 byte FIFO that we could enable

Reply:
Oh, nice